amptorch/preprocessing/utils.py

```diff
@@ -10,6 +13,92 @@
+
+        # fingerprints = normalize(fingerprints, norm='l2', axis=0)
+
+        if self.elementwise:
+            raise NotImplementedError
+
+        else:
+            if self.normalize:
+                mean = torch.mean(fingerprints, dim=0)
+                std = torch.std(fingerprints, dim=0, unbiased=False)
+                std[std < 1e-8] = 1
+                self.scale = {"offset": mean, "scale": std}
+                fingerprints_normalized = (
+                    fingerprints - self.scale["offset"]
+                ) / self.scale["scale"]
+                pca_reducer = PCA(n_components=self.num_pc).fit(
+                    fingerprints_normalized.numpy()
+                )
+            else:
+                pca_reducer = PCA(n_components=self.num_pc).fit(fingerprints.numpy())
+            self.pca_components = torch.tensor(
+                np.transpose(pca_reducer.components_), dtype=torch.get_default_dtype()
+            )
+            self.explained_variance = pca_reducer.explained_variance_
+            self.explained_variance_ratio = pca_reducer.explained_variance_ratio_
+            print(
+                "explained variance ratio: {}\n total: {}".format(
+                    self.explained_variance_ratio, np.sum(self.explained_variance_ratio)
+                )
+            )
+
+    def reduce(self, data_list, disable_tqdm=False):
+        if self.elementwise:
+            raise NotImplementedError
+
+        else:
+            for data in tqdm(
+                data_list,
+                desc="PCA reducing to: {} components".format(self.num_pc),
+                total=len(data_list),
+                unit=" images",
+                disable=disable_tqdm,
+            ):
+                fingerprint = data.fingerprint
+                # print("size before: {}".format(fingerprint.size()))
+                if self.normalize:
+                    fingerprint = (fingerprint - self.scale["offset"]) / self.scale[
+                        "scale"
+                    ]
+                fingerprint = torch.matmul(fingerprint, self.pca_components)
+                # print("size after: {}".format(fingerprint.size()))
+                data.fingerprint = fingerprint
+
+                if self.forcetraining:
+                    raise NotImplementedError
+                    # fp_idx = data.fprimes._indices()[0]
+                    # fp_idx_to_scale = fp_idx % data.fingerprint.shape[1]
+                    # _values = data.fprimes._values()
+
+                    # scale = self.scale["scale"][fp_idx_to_scale]
+                    # if self.transform == "standardize":
+                    #     _values /= scale
+                    # else:
+                    #     _values *= scale
+                    # _indices = data.fprimes._indices()
+                    # _size = data.fprimes.size()
+                    # data.fprimes = torch.sparse.FloatTensor(_indices, _values, _size)
+
+        return data_list
+
+
 class FeatureScaler:
```

Comment:
This looks fine - we should probably think of a cleaner way to do this or organize it. Not important for this PR, but something to keep in mind.